compiler/rustc_query_system/src/query/plumbing.rs

```diff
@@ -191,94 +215,97 @@ where
+        let id = lock.jobs.checked_add(1).unwrap();
+        let id = QueryShardJobId(NonZeroU32::new(id).unwrap());
 
-                let global_id = QueryJobId::new(id, shard, query.dep_kind);
+        let global_id = QueryJobId::new(id, shard, query.dep_kind);
 
-                let job = tcx.current_query_job();
-                let job = QueryJob::new(id, span, job);
+        let job = tcx.current_query_job();
+        let job = QueryJob::new(id, span, job);
 
-                entry.insert(QueryResult::Started(job));
+        // We optimize here for the happy path - that is, that the query will
+        // start without any cycles. So directly insert, and then if we do
+        // encounter 'unexpected' in the old result, fix things up.
+        match lock.active.insert(key.clone(), QueryResult::Started(job)) {
+            None => {
+                lock.jobs = id.0.get();
 
                 let owner = JobOwner { state, cache, id: global_id, key: (*key).clone() };
                 return TryGetJob::NotYetStarted(owner);
             }
-        };
-        mem::drop(state_lock);
-
-        // If we are single-threaded we know that we have cycle error,
-        // so we just return the error.
-        #[cfg(not(parallel_compiler))]
-        return TryGetJob::Cycle(cold_path(|| {
-            let error: CycleError = latch.find_cycle_in_stack(
-                tcx.try_collect_active_jobs().unwrap(),
-                &tcx.current_query_job(),
-                span,
-            );
-            let error = report_cycle(tcx.dep_context().sess(), error);
-            let value = query.handle_cycle_error(tcx, error);
-            cache.cache.store_nocache(value)
-        }));
-
-        // With parallel queries we might just have to wait on some other
-        // thread.
-        #[cfg(parallel_compiler)]
-        {
-            let result = latch.wait_on(tcx.current_query_job(), span);
-
-            if let Err(cycle) = result {
-                let cycle = report_cycle(tcx.dep_context().sess(), cycle);
-                let value = query.handle_cycle_error(tcx, cycle);
-                let value = cache.cache.store_nocache(value);
-                return TryGetJob::Cycle(value);
+            Some(QueryResult::Poisoned) => {
+                // put back poison, woops
+                lock.active.insert(key.clone(), QueryResult::Poisoned);
+                FatalError.raise();
             }
-
-            let cached = cache
-                .cache
-                .lookup(cache, &key, |value, index| {
-                    if unlikely!(tcx.dep_context().profiler().enabled()) {
-                        tcx.dep_context().profiler().query_cache_hit(index.into());
-                    }
-                    #[cfg(debug_assertions)]
-                    {
-                        cache.cache_hits.fetch_add(1, Ordering::Relaxed);
+            #[cfg_attr(not(parallel_compiler), allow(unused_mut))]
+            Some(QueryResult::Started(mut job)) => {
```

Comment:
Most of the code inside this arm depends on parallel vs non-parallel.
It may be simpler to have two arms: one for parallel, one for non-parallel.